astropy/io/fits/src/compressionmodule.c

```diff
@@ -386,18 +408,30 @@ void tcolumns_from_header(fitsfile* fileptr, PyObject* header,
     long long totalwidth;
     int status = 0;
 
-    get_header_int(header, "TFIELDS", &tfields, 0);
+    if (get_header_int(header, "TFIELDS", &tfields, 0, HDR_FAIL_VAL_NEGATIVE) == -1) {
+        return;
+    }
+    /* To avoid issues in the loop we need to limit the number of TFIELDs to
+       999. Otherwise we would exceed the maximum length of the keyword name of
+       8. This could lead to multiple accesses of the same header keyword with
+       snprintf because we limit it to 8 characters + null-termination. */
+    if (tfields > 999) {
+        PyErr_SetString(PyExc_ValueError, "The TFIELDS value exceeds 999.");
+        return;
+    }
 
     // This used to use PyMem_New, but don't do that; CFITSIO will later
     // free() this object when the file is closed, so just use malloc here
     // *columns = column = PyMem_New(tcolumn, (size_t) tfields);
     *columns = column = calloc((size_t) tfields, sizeof(tcolumn));
     if (column == NULL) {
+        PyErr_SetString(PyExc_MemoryError,
+                        "Couldn't allocate memory for columns.");
         return;
     }
 
 
-    for (idx = 1; idx <= tfields; idx++, column++) {
+    for (int idx = 1; idx <= tfields; idx++, column++) {
```

Comment:
Sorry, this is my fault but I guess we're compiling with (way) older versions of `gcc` on travis that don't recognize `c99` mode by default. We could build extensions with `--std=c99` but that's probably better left to a separate PR. Sorry to make you have to revert for now 😞